### DISCUSSION TITLE
Lightweight access tokens

### DISCUSSION THREAD

**Discussion by stianst** [keycloak/keycloak#9713] [Ideas]
From time to time there are use-cases that require something different to fully self-contained non-opaque access tokens. These include:

* Sensitive information: not including sensitive information in the access token
* Token size: due to number of claims, including roles or groups, the tokens are too large to be transmitted in HTTP headers

These can be solved with encrypted access tokens and/or reference tokens. However, these are both relatively complex:

* Encrypted access tokens are hard to use due to the need to deal with encryption keys
* Reference tokens are expensive to handle as these require a separate cache to be able to exchange a reference for the token

An alternative to the above could be lightweight access tokens. A lightweight access token is a token that contains limited claims, and the information needed to be able to generate a full access token. This allows the token to be a happy middle ground between a opaque vs non-opaque token and a encrypted vs a signed token.

The general idea on how to support lightweight access tokens is to add an options to all protocol mappers to add claims to the access token or not, then a lightweight access token can be defined by controlling what claims are available in the access token, and available in userinfo or token introspection endpoints.

Examples of how to use:

* Roles are not available in the access token, but only in the token introspection endpoint
* Personal user information is not available in the access token, but only in the userinfo endpoint. This raises one question on what clients can invoke the userinfo endpoint

One additional thing to consider here is who can invoke the token introspection and userinfo endpoint. The userinfo endpoint is not useful in cases of sensitive information as it can be invoked by anyone with a valid access token. The token introspection endpoint should only be possible to invoked by a confidential client that is included in the token audience.

To make it easier to use we may also want to be able to control how claims are included at the client scope level rather than on individual protocol mappers, as it can be error prone and complicated to update all protocol mappers for a client scope.

**Comment by stianst** [comment]
It seems we already have some support for this, but not sure if it fully works for all mappers/claims.

**Comment by mposolda** [comment]
@stianst Nice proposal. Few points and questions:

- If I understand correctly, the "Lightweight access token" can be opaque token readable only by Keycloak server. Keycloak server will need to understand token when sent to introspection-endpoint or user-info-endpoint. But client or resource-servers don't need to understand it right? Hence it can be JWT encrypted by Keycloak symmetric key? Or not even JWT in theory?

- How to control if client should issue "normal" JWT access tokens or lightweight access token? Do we want some client switch for this?

- +1 for control per protocol mapper if claims should be returned in the introspection response or not. I think that normal claims (like email, given_name etc) does not need to be returned as they are returned from UserInfo endpoint. So maybe just stuff like roles or web origins may be returned by default?

- I wonder that introspection endpoint can return different response for exactly same Lightweight access token based on which resource-server calls it? For example assume that there is Lightweight token LT1 issued for client "client1" .
  - Client1 sends the LT1 to the resource-server1. The resource-server1 will call introspection-endpoint with LT1 and Keycloak will return response with 10 roles relevant for resource-server1.
  - Client2 sends the LT2 to the resource-server2. The resource-server2 will call introspection-endpoint with LT2 and Keycloak will return response with 20 roles relevant for resource-server2 (Can be completely different roles than for resource-server1).

Does your proposal count with this option? Question is how to control what should be displayed for particular resource? For example for roles, it may be fine to count just with realm roles and/or client roles of the particular resource server?

I can imagine the last point can also provide solution to the admin console issue with big number of realms (WhoAmI) among other things.

**Reply by stianst** [reply]
The lightweight access token could (and probably should) still be non-opaque, as that makes it possible to read some claims directly from the token. I would also argue that from an implementation perspective it's easier to start with a non-opaque standard JWT.

Configuring lightweight access tokens is not a config option on a client directly, but rather you configure what client scopes / protocol mappers are added to the access token, and which are not.

As a separate thing we can add support for encrypted access tokens, and then the lightweight access tokens can be combined with encrypted access tokens.

**Reply by stianst** [reply]
The introspection endpoint response should not differ based on what client is invoking it. In the example you're giving actually the roles that go into the client is defined by client1, not by LT1 or LT2, and regardless which resource server invokes the introspection endpoint it should return the same result. To have different results client1 should really have different tokens for the different resource servers, or token exchange should be used.

**Reply by mposolda** [reply]
> The lightweight access token could (and probably should) still be non-opaque, as that makes it possible to read some claims directly from the token. I would also argue that from an implementation perspective it's easier to start with a non-opaque standard JWT.
> 
> Configuring lightweight access tokens is not a config option on a client directly, but rather you configure what client scopes / protocol mappers are added to the access token, and which are not.

Ok. IMO protocol mappers suits better to this than client scopes as they already have similar flags available (EG. `Add to ID Token`, `Add to Access Token`, `Add to User Info` ). So there can be probably just another similar switch like `Add to Introspection` ? 

> 
> As a separate thing we can add support for encrypted access tokens, and then the lightweight access tokens can be combined with encrypted access tokens.

Encrypted access tokens fits well into this IMO as nor the client or resource server needs to be able to read lightweight access token directly. Client has info about user claims from ID Token. Resource server can invoke either UserInfo or Introspection endpoint to receive the additional info (user claims or stuff like roles/web-origins). But +1 to consider encrypted access tokens as a follow-up.

**Reply by stianst** [reply]
Yeah, so first pass should be adding option to protocol mapper for sure, but was thinking perhaps a config option on a client scope would make sense as well.

**Reply by stianst** [reply]
I agree with encrypted tokens fits nicely into this, but can imagine that folks would want:

1. Lightweight, but still non opaque
2. Lightweight opaque (aka reference token)

Simplest to start with 1 ;)

**Comment by stianst** [comment]
There's a draft PR around a similar thing here:
https://github.com/keycloak/keycloak/pull/8914

**Comment by predictron-cloud** [comment]
Are there any news regarding the implementation of the lightweight token?

**Reply by jonkoops** [reply]
Please refrain from bumping discussions with comments that do not provide more insights related to the actual discussion.

**Comment by y-luis-rojo** [comment]
Any news on this?

**Reply by jonkoops** [reply]
Please refrain from bumping discussions with comments that do not provide more insights related to the actual discussion.

**Comment by tnorimat** [comment]
https://github.com/keycloak/keycloak/pull/22148 is also a draft PR.

**Comment by alexted** [comment]
Could you tell me, is this already available for use in production?

**Reply by agreedSkiing** [reply]
Seems to have been released in version 24
https://www.keycloak.org/2024/03/keycloak-2400-released#_lightweight_access_tokens_support
